Add driver for SAM4L's USBC and generic USB lib

daniel-scs opened the pull request:
This is still experimental, and provides just enough support so that
the SAM4L can be enumerated as a device (at least on Linux).

In the boot code for imix, the kernel attaches to the USB bus and
registers handlers to respond to host requests.

This commit includes:

- The beginnings of a generic interface to USB controllers
  in kernel/src/hil
- A driver (implementing the hil-controller interface) for the
  SAM4L's USBC in chips/sam4l
- A library in capsules/src/usb.rs that provides datatypes and
  serialization routines for speaking USB
- A client (implementing the hil-client interface) in
  capsules/src/usb_simple.rs that does all the platform-independent
  logic necessary for enumeration, like responding to host requests for
  various descriptors
- Some unsafe support code in kernel/src/common, most of which
  is not pretty and can probably be removed before mainlining

There is no user-space interface yet.

bradjc commented:
I just tried this on Hail and got the Zorpinator to show up (in my Linux VM on a Windows machine)! I did have to change the Hail clock setup to be:

    sam4l::pm::setup_system_clock(sam4l::pm::SystemClockSource::DfllRc32k, 48000000);

Hopefully there is something we can do in case a platform uses a different clock.

Full diff if anyone else wants to try:
```diff
diff --git a/boards/hail/src/main.rs b/boards/hail/src/main.rs
index bd19665d..bd6ea5ad 100644
--- a/boards/hail/src/main.rs
+++ b/boards/hail/src/main.rs
@@ -19,6 +19,7 @@ use kernel::{Chip, Platform};
 use kernel::hil;
 use kernel::hil::Controller;
 use kernel::hil::spi::SpiMaster;
+use kernel::hil::usb::Client;
 use kernel::mpu::MPU;
 use sam4l::usart;
 
@@ -86,6 +87,7 @@ struct Hail {
     rng: &'static capsules::rng::SimpleRng<'static, sam4l::trng::Trng<'static>>,
     ipc: kernel::ipc::IPC,
     crc: &'static capsules::crc::Crc<'static, sam4l::crccu::Crccu<'static>>,
+    usb: &'static capsules::usb_simple::SimpleClient<'static, sam4l::usbc::Usbc<'static>>,
 }
 
 impl Platform for Hail {
@@ -149,9 +151,12 @@ unsafe fn set_pin_primary_functions() {
     // PA[22].configure(None); // D2
     // PA[23].configure(Some(B)); // D4 - TWIMS0 SDA
     // PA[24].configure(Some(B)); // D5 - TWIMS0 SCL
-    // UART Mode
-    PA[25].configure(Some(B)); // RX - USART2 RXD
-    PA[26].configure(Some(B)); // TX - USART2 TXD
+    // // UART Mode
+    // PA[25].configure(Some(B)); // RX - USART2 RXD
+    // PA[26].configure(Some(B)); // TX - USART2 TXD
+    // USB Mode
+    PA[25].configure(Some(A)); // RX - USB D-
+    PA[26].configure(Some(A)); // TX - USB D+
 
     PB[00].configure(Some(A)); // SENSORS_SDA - TWIMS1 SDA
     PB[01].configure(Some(A)); // SENSORS_SCL - TWIMS1 SCL
@@ -175,8 +180,9 @@ unsafe fn set_pin_primary_functions() {
 pub unsafe fn reset_handler() {
     sam4l::init();
 
-    sam4l::pm::setup_system_clock(sam4l::pm::SystemClockSource::ExternalOscillatorPll,
-                                  48000000);
+    // sam4l::pm::setup_system_clock(sam4l::pm::SystemClockSource::ExternalOscillatorPll,
+    //                               48000000);
+    sam4l::pm::setup_system_clock(sam4l::pm::SystemClockSource::DfllRc32k, 48000000);
 
     // Source 32Khz and 1Khz clocks from RC23K (SAM4L Datasheet 11.6.8)
     sam4l::bpm::set_ck32source(sam4l::bpm::CK32Source::RC32K);
@@ -378,6 +384,13 @@ pub unsafe fn reset_handler() {
         128/8);
     sam4l::crccu::CRCCU.set_client(crc);
 
+    // USB
+    let usb_client = static_init!(
+        capsules::usb_simple::SimpleClient<'static, sam4l::usbc::Usbc<'static>>,
+        capsules::usb_simple::SimpleClient::new(&sam4l::usbc::USBC),
+        288/8);
+    sam4l::usbc::USBC.set_client(usb_client);
+
 
     let hail = Hail {
         console: console,
@@ -394,6 +407,7 @@ pub unsafe fn reset_handler() {
         rng: rng,
         ipc: kernel::ipc::IPC::new(),
         crc: crc,
+        usb: usb_client,
     };
 
     // Need to reset the nRF on boot
@@ -415,6 +429,10 @@ pub unsafe fn reset_handler() {
     let mut chip = sam4l::chip::Sam4l::new();
     chip.mpu().enable_mpu();
 
+    // Test USB
+    hail.usb.enable();
+    hail.usb.attach();
+
     // Uncomment to measure overheads for TakeCell and MapCell:
     // test_take_map_cell::test_take_map_cell();
```

daniel-scs commented:
Oh cool, thanks for trying it out Brad!

I was piggybacking on the main clock currently used for `imix`, but yes of course we should have a way to get it working with arbitrary platforms.  It looks like that's going to require some new logic for clock selection.

alevy commented:
Works well for me as well. @daniel-scs functionality-wise this is definitely ready to be merged (in the sense that just enumerating worth merging).

You ended up using a non-standard idioms, so I'll go through and ask questions about those, though they may be perfectly fine to use, especially since everything except the `simple_usb` capsule and the HIL is internal to the implementation. I'll try to be explicit in the comments about which things are requests for changes vs just questions for clarification.

alevy commented on capsules/src/usb.rs:
```diff
@@ -0,0 +1,467 @@
+    fn write_to(&self, buf: &mut [u8]) -> usize;
+}
+
+pub struct DeviceDescriptor {
```
Comments in general for these structure would be good, but especially for this one and it's fields which are somewhat obscure if you don't know the USB protocol. Even just adapting the basic descriptions from the [USB Made Simple](http://www.usbmadesimple.co.uk/) would be enough I think:

![screenshot](https://cloud.githubusercontent.com/assets/4689/26756659/ce2abd34-4875-11e7-8001-99bcf0b8f7d2.png)

alevy commented on capsules/src/usb.rs:
```diff
@@ -0,0 +1,467 @@
+impl Descriptor for InterfaceDescriptor {
+    fn size(&self) -> usize { 9 }
+
+    fn write_to(&self, buf: &mut [u8]) -> usize {
```
should bounds-check here. since it's only 9 bytes, it might be simplest to create the buffer on the stack with the expected size and use a fold or a built in copy function or something (might even get inlined out... who knows...)

alevy commented on capsules/src/usb.rs:
```diff
@@ -0,0 +1,467 @@
+    }
+
+    // Encode as utf16-le
+    fn write_to(&self, buf: &mut [u8]) -> usize {
```
bounds check

alevy commented on capsules/src/usb.rs:
```diff
@@ -0,0 +1,467 @@
+        2 + (2 * self.langs.len())
+    }
+
+    fn write_to(&self, buf: &mut [u8]) -> usize {
```
bounds-check

alevy commented on capsules/src/usb.rs:
```diff
@@ -0,0 +1,467 @@
+impl Descriptor for ConfigurationDescriptor {
+    fn size(&self) -> usize { 9 }
+
+    fn write_to(&self, buf: &mut [u8]) -> usize {
```
bounds-check

alevy commented on capsules/src/usb.rs:
```diff
@@ -0,0 +1,467 @@
+impl Descriptor for DeviceDescriptor {
+    fn size(&self) -> usize { 18 }
+
+    fn write_to(&self, b: &mut [u8]) -> usize {
```
bounds-check

alevy commented on chips/sam4l/src/usbc/common_register.rs:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    #[inline]
+    pub fn clr_bit(self, bit_index: u32) {
```
`clear_bit`

alevy commented on chips/sam4l/src/usbc/common_register.rs:
```diff
@@ -0,0 +1,244 @@
+/// A memory-mapped register
+#[derive(Copy, Clone)]
+pub struct Reg<T> {
+    addr: *mut u32,
```
Why not just `*mut T`. Then you don't need the `PhantomData` field or any of the `ToWord/FromWord` stuff

alevy commented on chips/sam4l/src/usbc/common_register.rs:
```diff
@@ -0,0 +1,244 @@
+#![allow(dead_code)]
```
This seems a bit like re-inventing `VolatileCell`. Can you explain the purpose of creating a new type rather than extending `VolatileCell` with the extra functionality?

alevy commented on chips/sam4l/src/usbc/common_register.rs:
```diff
@@ -0,0 +1,244 @@
+    }
+}
+
+/// A write-only memory-mapped register, where any zero bits written have no effect
```
I don't think "where any zero bits written have no effect" is important here. The type seems equally functional for write-only registers where 0-bits have an effect and for those that don't.

alevy commented on chips/sam4l/src/usbc/common_register.rs:
```diff
@@ -0,0 +1,244 @@
+}
+
+/// An array of memory-mapped registers
+pub struct Regs<T> {
```
Aren't these plural types (`Regs`, `RegsW`, `RegsR`) extraneous since you can just use a fixed-size array? e.g. `[RegR; 22]`? It compiles to the same thing (fixed-sized arrays don't have a length field since the length is deduced for bounds checking etc at compile-time).

alevy commented on chips/sam4l/src/usbc/common_register.rs:
```diff
@@ -0,0 +1,244 @@
+        Regs { addr: addr, phantom: PhantomData }
+    }
+
+    pub fn n(&self, index: u32) -> Reg<T> {
```
I think this is super duper unsafe. There isn't anything preventing me from reading/writing arbitrary memory by calling `myregs_field.n(0xdeadbeef)` and getting back a seemingly valid `Reg<u32>`.

alevy commented on chips/sam4l/src/usbc/common_register.rs:
```diff
@@ -0,0 +1,244 @@
+        RegsW { addr: addr, phantom: PhantomData }
+    }
+
+    pub fn n(&self, index: u32) -> RegW<T> {
```
ditto on unsafety

alevy commented on chips/sam4l/src/usbc/common_register.rs:
```diff
@@ -0,0 +1,244 @@
+}
+
+impl<T: ToWord> BitField<T> {
+    pub const fn new(reg: Reg<u32>, shift: u32, bits: u32) -> Self {
```
hrm.... what if `size_of::<T>() + shift > `size_of::<u32>`?

alevy replied:
If nothing else, the constructors should at least be marked unsafe.

alevy commented on chips/sam4l/src/usbc/common_register.rs:
```diff
@@ -0,0 +1,244 @@
+}
+
+/// A bitfield of a memory-mapped register
+pub struct BitField<T> {
```
Is the reason for this type basically that Rust doesn't have support for non-byte-aligned values in structs, and this lets you avoid doing binary arithmetic in that case? If so, I get it, but it seems _very_ dangerous to me from a type-safety perspective. _Perhaps_ if it only supported integral fields or something, but even then, we would have to think very hard about the implications...

alevy commented on chips/sam4l/src/usbc/register_macros.rs:
```diff
@@ -0,0 +1,65 @@
+//! Macros for defining USBC registers
```
These are wonderful macros! Most the comments I would have here are really about my comments in the common_register module.

alevy commented on chips/sam4l/src/usbc/register_macros.rs:
```diff
@@ -0,0 +1,65 @@
+macro_rules! reg {
+    [ $offset:expr, $description:expr, $name:ident, "RW" ] => {
+        #[allow(dead_code)]
+        pub const $name: Reg<u32> = Reg::new((USBC_BASE + $offset) as *mut u32);
```
`Reg` et al should be fully qualified here (that's generally true in macros) otherwise it would require the user of the macro to have it in scope already.

alevy commented on kernel/src/common/copy_slice.rs:
```diff
@@ -0,0 +1,50 @@
+//! A wrapper around an immutable slice of `Copy` elements
+//! that provides access via a mutable slice
```
As it's written, I'm skeptical this is safe. Here's one example:

```rust
#[derive(Copy,Clone)]
enum Foo {
  Num(usize),
  Ref(&'a usize)
}

let myslice: CopySlice<Foo> = ...;

let Ref(x) = myslice.as_slice()[0]; // `x` is a `&'static usize`

myslice.as_mut()[0] = Num(0xdeadbeef);

// now `x` is actually a reference to `0xdeadbeef`. oops...
```

However, I _think_ what you're trying to get at is a (more ergonomic) version of something like `&[Cell<T>]`, which can be shared (it's not a mutable reference) and each element can be modified since elements are `Cell`s, but the index operation gives you a `Cell<T>` rather a `T`, so you can never get multiple references to the same actual memory of an element:

```rust
myslice[0].get();
myslice[0].set(1234);
```

This _may_ be sufficient for what you care about in which case, despite the slightly less ergonomic type it might be better to just use `&[Cell<T>]`, or you may be able to write a similar wrapper type that implements the same semantics and hides some of the details behind the `Index` and `IndexMut` traits.

alevy commented on kernel/src/common/copy_slice.rs:
```diff
@@ -0,0 +1,50 @@
+}
+
+pub fn static_mut_bytes_8() -> &'static mut [u8] {
+    unsafe {
```
I think this will only do anything reasonable if it's only ever called once, in which case the function should really be marked unsafe. Basically, the statically allocated slice is always the same one.

alevy commented on kernel/src/common/copy_slice.rs:
```diff
@@ -0,0 +1,50 @@
+}
+
+pub fn static_mut_bytes_100() -> &'static mut [u8] {
+    unsafe {
```
ditto

alevy commented on kernel/src/common/static_fmt.rs:
```diff
@@ -0,0 +1,64 @@
+//! A macro static_fmt! that provides a small amount of static storage
+//! for formatting a string.  The value of the macro has type `&'static str`.
+//! Each invocation overwrites what was written previously.
```
Similar to `static_mut_bytes_8`, I think this will only work as expected if the resulting string is used and discarded before any other invocation to `static_fmt!` is called, which is risky.

I wonder if a better solution to error formatting is to just provide an enum with a bunch of cases, where constructors to the enum can take arguments needed for actually formatting a final string, then implement the `Display` trait for the enum.

alevy commented on kernel/src/common/volatile_slice.rs:
```diff
@@ -0,0 +1,52 @@
+use core::ptr;
```
What problem does this solve? Did you encounter an issue with the USB descriptors in memory or something?

alevy commented:
Whew! @daniel-scs this is a herculean effort. :1st_place_medal:

daniel-scs commented:
Woo, thanks so much for the detailed review!  I've addressed
most of it in this commit, and will save a few parts for the
merge-ready moment.

Thanks to your helpful comments, I was able to eliminate the
entirety of my unsafe-hacks dumping-ground in
kernel::common.  Mostly this was about fighting with Rust to
get buffers that didn't "optimize-out" I/O.  Your suggestion
of `[Cell<u8>]` and `[VolatileCell<u8>]` worked great, 
provided they are marked `mut`; otherwise even
`volatile_read()` just produces statically-calculated
values!  Maybe that makes sense somehow.

In terms of where to allocate the needed buffers, I was
grateful for your pointing out that the healthy approach is
to let the run-once function allocate them statically and
then pass them *into* the capsule.  (I should have noticed the
other modules that do that.)  Even better, for my case it
turned out to be possible to just allocate them in the
capsule datastructure itself and thus let `new()` manufacture
them arbitrarily-many times.  (I think I tried that at first
but gave up trying to make the lifetimes work.)

Your comments also helped me simplify the
register-manipulation datatypes and eliminate some unsafe
practices.  My goal with this stuff is just to make the
application-specific logic as clear and concise as possible,
and to avoid doing arithmetic when laying out the registers.
I think the result is that my `Reg` type is unsafe each time
you construct one (in this commit I've marked the 
constructors unsafe, as you suggest), because it allows you 
to read/write an arbitrary memory location; whereas the
usual approach has been to do *one* unsafe operation to map 
a large array of register-representation datatypes onto the
starting address of the appropriate memory-mapped region.
This probably means there are more opportunities to cause 
havoc with a typo in my approach, so I'll probably revert to
the existing custom.

As I've left it for now, it looks to me like the `Bitfield`
datatypes don't introduce any extra unsafety themselves:
it's just that they use the `Reg` types, whose construction 
is dangerous because they allow you to specify the address
you want to bit-twiddle.

Anyway, I think it makes sense to wait on a merge until I
have an actual application working: it sounds like just
setting up a pair of bulk endpoints is sufficient to get
basic serial IO working with Linux, so that's my plan.  I'll
address the remainder of your suggestions (file structure,
documentation comments, and bounds checks) while I do that.

Thanks again!

bradjc commented on chips/sam4l/src/usbc/mod.rs:
```diff
@@ -0,0 +1,874 @@
+use kernel::hil;
+use kernel::hil::usb::*;
+use kernel::common::take_cell::MapCell;
+use kernel::common::volatile_cell::VolatileCell;
```
`use kernel::common::VolatileCell;`

bradjc commented on kernel/src/hil/usb.rs:
```diff
@@ -0,0 +1,73 @@
+//! Interface to USB controller hardware
+
+use common::volatile_cell::*;
```
Is there any more than just `VolatileCell`?

daniel-scs commented:
The above addresses most prior concerns and is probably ok to merge (debugging hooks are removed from boot, so it shouldn't disturb anything), but I'm going to do the following things this week and it might make sense to wait on a review until they've arrived here:

- usb support for Hail
- command-line utility for piping data through a USB endpoint, so you can see that it all works

alevy commented:
I have a command line utility available _somewhere_ (perhaps in `tools/`) before merging, but not necessarily USB support for hail. The PR is already pretty big, and in any case, adding support for hail means requiring additional hardware, so not _that_ crucial for a first merge.

alevy commented on capsules/src/usb_user.rs:
```diff
@@ -0,0 +1,137 @@
+//! USB system call interface
```
Perhaps this (and the associated user-land library and app) can wait? I think if this PR implements a basic set of features for the SAM4L, and has an initial HIL that's already a fair bit. Along with the `usb_client` capsule, I think that'll be good.

daniel-scs replied:
Ok, I can cut out the syscalls and make that part of the next PR if you prefer.

And actually yes I think I'll leave off Hail because it may lead to interesting clock-management issues.

For the utility, I've got a simple one I can just drop in today that gets a few bytes into and out of the host, so I might as well pop that in.

daniel-scs commented:
Sooooo I realized the reason I wanted to bundle the very simple syscall stuff with this PR is that it allows you to actually test the functionality (by loading a trivial app) without adding some hack to the boot code.  So I've left it in for now.  But let me know if it is awkward to take on all of these bits at once.

Anyway after the last few commits I believe I've addressed all reviewer concerns and am excited to merge!

niklasad1 commented:
I think you just need to run uncrustify in userland to make Travis pass by: 
```$ make formatall```

daniel-scs commented:
Ok yeah let me just do the source-prettifying ...

bradjc commented on capsules/src/usb_user.rs:
```diff
@@ -0,0 +1,138 @@
+//!
+//! ## Instantiation
+//!
+//! _TODO_
```
Would be good to have this before merging, as with other new capsule PRs.

daniel-scs replied:
Good idea, done!
